### DISCUSSION TITLE
TestCaseSource

### DISCUSSION THREAD

**Discussion by guy1407** [nunit/nunit#4266] [Q&A] [answered]
Hello. 
I have two separate class files with tests.
One has testcase with testcasesource method.
One is normal. 
When I run the "normal test" with no test case - the method that gets the value for the test cases is executed! (I had a breakpoint there)
Is that "by design"?
What if I wish to have several testcasesource methods (for different tests)?

**Comment by manfred-brands** [comment, marked as answer]
The testcasesource is executed to determine all the testcases. You can see all them in VS TestExplorer.

**Reply by guy1407** [reply]
Many thanks for your quick reply.
Do you have a sample code, showing how to pass a parameter to the method that fetches data?

**Reply by manfred-brands** [reply]
There is an example in the [nunit documentation](https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcasesource.html)

**Reply by mikkelbu** [reply]
Note that this is the default design. For people with "expensive" testcasesources (which is in general discouraged) one can use prefilter - see e.g. https://docs.nunit.org/articles/vs-test-adapter/Tips-And-Tricks.html#prefilter for how to do this in VS - to only run a selection of the testscasesources/tests. Note that there have been several issues with this in edge-cases, so the functionality is disabled as default, but I think most of them have been solved by now.

Prefilters are also supported by NUnitLite - https://docs.nunit.org/articles/nunit/running-tests/NUnitLite-Options.html?q=prefilter - but I cannot remember if they ever reached NUnit-Console.

**Reply by stevenaw** [reply]
I don't think it surfaced in the console yet, but it looks to currently be in the 4.0 milestone
https://github.com/nunit/nunit-console/issues/438 is the issue
